File: app/src/main/java/com/duncan/dpi/activity/MainActivity.java
package com.duncan.dpi.activity;


import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.design.widget.TextInputLayout;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.EditText;
import android.widget.TextView;

import com.duncan.dpi.util.CalcUtil;
import com.duncan.dpi.model.Device;
import com.duncan.dpi.view.DynamicViewController;
import com.duncan.dpi.R;

public class MainActivity extends AppCompatActivity {
    Toolbar toolbar;
    DynamicViewController controller;
    TextInputLayout widthContainer, heightContainer, screenSizeContainer;
    EditText width, height, screenSize;
    TextView answer;
    final int DEVICE_LIST_REQUEST_CODE = 9999;
    boolean viewedDeviceListDialog = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //Vars
        toolbar = (Toolbar)findViewById(R.id.toolbar);

        this.controller = new DynamicViewController(findViewById(R.id.dynamicView));
        this.answer = (TextView)findViewById(R.id.answer);

        this.widthContainer = (TextInputLayout)findViewById(R.id.widthContainer);
        this.heightContainer = (TextInputLayout)findViewById(R.id.heightContainer);
        this.screenSizeContainer = (TextInputLayout)findViewById(R.id.screenSizeContainer);

        this.width = this.widthContainer.getEditText();
        this.height = this.heightContainer.getEditText();
        this.screenSize = this.screenSizeContainer.getEditText();

        //Setup
        setSupportActionBar(toolbar);
        /*this.widthContainer.setErrorEnabled(true);
        this.heightContainer.setErrorEnabled(true);
        this.screenSizeContainer.setErrorEnabled(true);*/

        setupTextWatchers();
    }

    /**
     * Setup the view_dynamic view to reflect texts as they are typed
     */
    private void setupTextWatchers() {
        TextWatcher attemptCalcOnTextChange = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                attemptCalculation();
            }

            @Override
            public void afterTextChanged(Editable s) {}
        };
        this.width.addTextChangedListener(attemptCalcOnTextChange);
        this.height.addTextChangedListener(attemptCalcOnTextChange);
        this.screenSize.addTextChangedListener(attemptCalcOnTextChange);
    }

    /**
     * Attempt to calculate DPI
     */
    private void attemptCalculation() {
        int w, h;
        double ss;
        try {
            w = Integer.parseInt(width.getText().toString());
            h = Integer.parseInt(height.getText().toString());
            ss = Double.parseDouble(screenSize.getText().toString());
            controller.setWidth(w);
            controller.setHeight(h);
            controller.setScreenSize(ss);
            answer.setText(CalcUtil.calculateDPI(w, h, ss) + " dpi");
        } catch (Exception e) {
            controller.setWidth(-1);
            controller.setHeight(-1);
            controller.setScreenSize(-1);
            answer.setText("- dpi");
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.deviceList) {
            if (!viewedDeviceListDialog) {
                AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this, R.style.DialogTheme);
                builder.setMessage(R.string.dialog_content)
                        .setTitle(R.string.device_list)
                        .setPositiveButton(R.string.dialog_agree, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                viewedDeviceListDialog = true;
                                Intent intent = new Intent(MainActivity.this, DeviceListActivity.class);
                                startActivityForResult(intent, DEVICE_LIST_REQUEST_CODE);
                            }
                        })
                        .setNegativeButton(R.string.dialog_cancel, null);
                AlertDialog dialog = builder.create();
                dialog.show();
            } else {
                //Since seen before, no need to show again.
                Intent intent = new Intent(MainActivity.this,  DeviceListActivity.class);
                startActivityForResult(intent, DEVICE_LIST_REQUEST_CODE);
            }
        }
        return false;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == DEVICE_LIST_REQUEST_CODE && resultCode == Activity.RESULT_OK) {
            Device device = data.getParcelableExtra("device");
            width.setText("" + device.getScreenWidth());
            height.setText("" + device.getScreenHeight());
            screenSize.setText(String.format("%.2f", device.getScreenSize()));
            attemptCalculation();
            Snackbar.make(width, String.format("%s %s", getString(R.string.message_populate), device.getTitle()), Snackbar.LENGTH_LONG)
                    .show();
            //Toast.makeText(MainActivity.this, "Inserted data from " + device.getTitle(), Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/duncan/dpi/activity/MainActivity.kt
package com.duncan.dpi.activity


import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import com.duncan.dpi.R
import com.duncan.dpi.databinding.ActivityMainBinding
import com.duncan.dpi.model.Device
import com.duncan.dpi.util.CalcUtil
import com.google.android.material.snackbar.Snackbar
import com.jakewharton.rxbinding.widget.RxTextView
import com.jakewharton.rxbinding.widget.TextViewTextChangeEvent
import rx.android.schedulers.AndroidSchedulers
import rx.functions.Action1
import rx.functions.Func1
import java.util.concurrent.TimeUnit

class MainActivity : AppCompatActivity() {
    internal val DEVICE_LIST_REQUEST_CODE = 9999
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val filter = Func1<TextViewTextChangeEvent, Boolean> { event -> event.text().isNotEmpty() }
        val action = Action1<TextViewTextChangeEvent> { event -> attemptCalculation() }
        RxTextView.textChangeEvents(binding.inputWidth.editText!!)
                .debounce(700, TimeUnit.MILLISECONDS)
                .filter(filter)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(action)
        RxTextView.textChangeEvents(binding.inputHeight.editText!!)
                .debounce(700, TimeUnit.MILLISECONDS)
                .filter(filter)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(action)
        RxTextView.textChangeEvents(binding.inputScreenSize.editText!!)
                .debounce(700, TimeUnit.MILLISECONDS)
                .filter(filter)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(action)
    }

    /**
     * Attempt to calculate DPI
     */
    private fun attemptCalculation() {
        val width = try { binding.inputWidth.editText?.text.toString().toInt() } catch (e: NumberFormatException) { 0 }
        val height = try { binding.inputHeight.editText?.text.toString().toInt() } catch (e: NumberFormatException) { 0 }
        val screenSize = try { binding.inputScreenSize.editText?.text.toString().toDouble() } catch (e: NumberFormatException) { 0.0 }
        binding.labelDensity.text = "${CalcUtil.calculateDPI(width, height, screenSize)} dpi"

        val screenWidth = resources.displayMetrics.widthPixels
        val screenHeight = resources.displayMetrics.heightPixels

        val maxWidth = screenWidth * 0.5
        val maxHeight = screenHeight * 0.6

        val params = binding.imageScreen.layoutParams

        // Attempt maxWidth
        params.width = maxWidth.toInt()
        val intendedHeight = maxWidth / width * height
        if (intendedHeight > maxHeight) {
            // Use maxWidth
            val intendedWidth = maxHeight / height * width
            params.width = intendedWidth.toInt()
            params.height = maxHeight.toInt()
        } else {
            params.height = intendedHeight.toInt()
        }

        binding.imageScreen.layoutParams = params
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putString("width", binding.inputWidth.editText?.text.toString())
        outState.putString("height", binding.inputHeight.editText?.text.toString())
        outState.putString("screenSize", binding.inputScreenSize.editText?.text.toString())
        outState.putString("result", binding.labelDensity.text.toString())
        super.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        binding.inputWidth.editText?.setText(savedInstanceState.getString("width"))
        binding.inputHeight.editText?.setText(savedInstanceState.getString("height"))
        binding.inputScreenSize.editText?.setText(savedInstanceState.getString("screenSize"))
        binding.labelDensity.text = savedInstanceState.getString("result")
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when(item.itemId) {
            R.id.deviceList -> {
                //Since seen before, no need to show again.
                val intent = Intent(this@MainActivity, DeviceListActivity::class.java)
                startActivityForResult(intent, DEVICE_LIST_REQUEST_CODE)
            }
        }
        return false
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == DEVICE_LIST_REQUEST_CODE && resultCode == Activity.RESULT_OK) {
            val device = data?.getParcelableExtra<Device>("device")
            binding.inputWidth.editText?.setText("${device?.width}")
            binding.inputHeight.editText?.setText("${device?.height}")
            binding.inputScreenSize.editText?.setText(String.format("%.2f", device?.screenSize))
            attemptCalculation()
            Snackbar.make(binding.imageScreen, String.format("%s %s", getString(R.string.message_populate), device?.name), Snackbar.LENGTH_LONG)
                    .show()
        }
    }
}
